asf_heat_pump_suitability/pipeline/prepare_features/property_density.py

```diff
@@ -2,67 +2,76 @@
 from asf_heat_pump_suitability.getters import get_datasets
 
 
-def extend_df_with_property_density(enhanced_epc_df: pl.DataFrame) -> pl.DataFrame:
+def generate_df_property_density() -> pl.DataFrame:
     """
-    Add property density feature to the EPC dataset
-    Args:
-        enhanced_epc_df (pl.DataFrame): EPC dataset with additional features
+    Generate dataframe with property density (households per km2) per LSOA/2011 Data Zone in England and Wales, and Scotland,
+    respectively, using Standard Area Measurement of ‘Land Area’ (Area to Mean High Water Excluding Area of Inland Water).
+
     Returns:
-        pl.DataFrame: EPC dataset with property density feature
+        pl.DataFrame: households per km2 per LSOA/Data Zone in England, Wales, and Scotland
     """
-    # Fill None values with a default value (e.g., 0)
-    enhanced_epc_df = enhanced_epc_df.with_columns(
-        [
-            pl.col("Number of households 2021")
-            .fill_null(0)
-            .alias("Number of households 2021"),
-            pl.col("Land Count (Area in KM2)")
-            .fill_null(0)
-            .alias("Land Count (Area in KM2)"),  # Fill with 1 to avoid division by zero
```

Comment:
It's no longer needed because when calculating property density, instead of calculating for the whole EPC dataset, I only calculate it where sufficient data is available - i.e. both household count and land area data is available. Household counts and area data is joined with an `inner` join. The minimum value for land area is 0.0097 km2, so we don't end up dividing anything by 0.

The resulting property density df is then left joined to the full EPC dataset. So where property density information is not available, there will be nulls in the final EPC dataset in this column.